File: third-party/thrift/src/thrift/lib/javadeprecated/src/main/java/com/facebook/thrift/protocol/THeaderProtocol.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.thrift.protocol;

import com.facebook.thrift.TApplicationException;
import com.facebook.thrift.TException;
import com.facebook.thrift.transport.THeaderException;
import com.facebook.thrift.transport.THeaderTransport;
import com.facebook.thrift.transport.TTransport;
import java.util.List;
import java.util.Map;

/** Pass through protocol for use with THeaderTransport servers */
public class THeaderProtocol extends TProtocol {

  private TProtocol proto;
  private int protoId;

  private void resetProtocol() throws TProtocolException {
    // We guarantee trans_ to be of type THeaderTransport, parent class does not
    if (proto != null && protoId == ((THeaderTransport) trans_).getProtocolId()) {
      return;
    }

    protoId = ((THeaderTransport) trans_).getProtocolId();
    switch (protoId) {
      case THeaderTransport.T_BINARY_PROTOCOL:
        proto = new TBinaryProtocol(trans_, true, true);
        break;
      case THeaderTransport.T_COMPACT_PROTOCOL:
        proto = new TCompactProtocol(trans_);
        break;
      default:
        throw new TProtocolException("Unknown protocol id: " + protoId);
    }
  }

  /** Factory */
  @SuppressWarnings("serial")
  public static class Factory implements TProtocolFactory {
    List<THeaderTransport.ClientTypes> clientTypes;

    public Factory(List<THeaderTransport.ClientTypes> clientTypes) {
      this.clientTypes = clientTypes;
    }

    public Factory() {}

    public TProtocol getProtocol(TTransport trans) {
      if (trans instanceof THeaderTransport) {
        // Make sure we call the THeaderTransport specific constructor if
        // we can.
        return new THeaderProtocol((THeaderTransport) trans);
      } else {
        return new THeaderProtocol(trans, clientTypes);
      }
    }
  }

  /** Constructor */
  public THeaderProtocol(TTransport trans, List<THeaderTransport.ClientTypes> clientTypes) {
    this(new THeaderTransport(trans, clientTypes));
  }

  public THeaderProtocol(THeaderTransport trans) {
    super(trans);
    try {
      resetProtocol();
    } catch (TProtocolException tpe) {
      // This actually can't occur here - always constructed with a
      // TBinaryProtocol.
      throw new RuntimeException(tpe);
    }
  }

  @Override
  public void writeMessageBegin(TMessage message) throws TException {
    proto.writeMessageBegin(message);
  }

  @Override
  public void writeMessageEnd() throws TException {
    proto.writeMessageEnd();
  }

  @Override
  public void writeStructBegin(TStruct struct) throws TException {
    proto.writeStructBegin(struct);
  }

  @Override
  public void writeStructEnd() throws TException {
    proto.writeStructEnd();
  }

  @Override
  public void writeFieldBegin(TField field) throws TException {
    proto.writeFieldBegin(field);
  }

  @Override
  public void writeFieldEnd() throws TException {
    proto.writeFieldEnd();
  }

  @Override
  public void writeFieldStop() throws TException {
    proto.writeFieldStop();
  }

  @Override
  public void writeMapBegin(TMap map) throws TException {
    proto.writeMapBegin(map);
  }

  @Override
  public void writeMapEnd() throws TException {
    proto.writeMapEnd();
  }

  @Override
  public void writeListBegin(TList list) throws TException {
    proto.writeListBegin(list);
  }

  @Override
  public void writeListEnd() throws TException {
    proto.writeListEnd();
  }

  @Override
  public void writeSetBegin(TSet set) throws TException {
    proto.writeSetBegin(set);
  }

  @Override
  public void writeSetEnd() throws TException {
    proto.writeSetEnd();
  }

  @Override
  public void writeBool(boolean b) throws TException {
    proto.writeBool(b);
  }

  @Override
  public void writeByte(byte b) throws TException {
    proto.writeByte(b);
  }

  @Override
  public void writeI16(short i16) throws TException {
    proto.writeI16(i16);
  }

  @Override
  public void writeI32(int i32) throws TException {
    proto.writeI32(i32);
  }

  @Override
  public void writeI64(long i64) throws TException {
    proto.writeI64(i64);
  }

  @Override
  public void writeDouble(double dub) throws TException {
    proto.writeDouble(dub);
  }

  @Override
  public void writeFloat(float flt) throws TException {
    proto.writeFloat(flt);
  }

  @Override
  public void writeString(String str) throws TException {
    proto.writeString(str);
  }

  @Override
  public void writeBinary(byte[] bin) throws TException {
    proto.writeBinary(bin);
  }

  /** Reading methods. */

  /** Helper method to throw an error back to the endpoint */
  public void notifyEndpoint(String msg) throws TException {
    if (proto != null) {
      writeMessageBegin(new TMessage("", TMessageType.EXCEPTION, 0));
      TApplicationException ex = new TApplicationException(msg);
      ex.write(this);
      writeMessageEnd();
      trans_.flush();
    }
  }

  @Override
  public TMessage readMessageBegin() throws TException {
    // Read the next frame, and change protocols if needed
    try {
      ((THeaderTransport) trans_)._resetProtocol();
      resetProtocol();
    } catch (THeaderException e) {
      // THeaderExceptions are exceptions we want thrown back to the endpoint
      // Such as unknown transforms or protocols.  Endpoint may retry
      notifyEndpoint(e.toString());
    }
    return proto.readMessageBegin();
  }

  @Override
  public void readMessageEnd() throws TException {
    proto.readMessageEnd();
  }

  public TStruct readStructBegin(
      Map<Integer, com.facebook.thrift.meta_data.FieldMetaData> metaDataMap) throws TException {
    return proto.readStructBegin(metaDataMap);
  }

  @Override
  public void readStructEnd() throws TException {
    proto.readStructEnd();
  }

  @Override
  public TField readFieldBegin() throws TException {
    return proto.readFieldBegin();
  }

  @Override
  public void readFieldEnd() throws TException {
    proto.readFieldEnd();
  }

  @Override
  public TMap readMapBegin() throws TException {
    return proto.readMapBegin();
  }

  @Override
  public void readMapEnd() throws TException {
    proto.readMapEnd();
  }

  @Override
  public TList readListBegin() throws TException {
    return proto.readListBegin();
  }

  @Override
  public void readListEnd() throws TException {
    proto.readListEnd();
  }

  @Override
  public TSet readSetBegin() throws TException {
    return proto.readSetBegin();
  }

  @Override
  public void readSetEnd() throws TException {
    proto.readSetEnd();
  }

  @Override
  public boolean readBool() throws TException {
    return proto.readBool();
  }

  @Override
  public byte readByte() throws TException {
    return proto.readByte();
  }

  @Override
  public short readI16() throws TException {
    return proto.readI16();
  }

  @Override
  public int readI32() throws TException {
    return proto.readI32();
  }

  @Override
  public long readI64() throws TException {
    return proto.readI64();
  }

  @Override
  public double readDouble() throws TException {
    return proto.readDouble();
  }

  @Override
  public float readFloat() throws TException {
    return proto.readFloat();
  }

  @Override
  public String readString() throws TException {
    return proto.readString();
  }

  @Override
  public byte[] readBinary() throws TException {
    return proto.readBinary();
  }
}

File: third-party/thrift/src/thrift/lib/py/protocol/THeaderProtocol.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# pyre-unsafe

from __future__ import absolute_import, division, print_function, unicode_literals

from thrift.protocol.TBinaryProtocol import TBinaryProtocolAccelerated
from thrift.protocol.TCompactProtocol import TCompactProtocolAccelerated
from thrift.protocol.TProtocol import TProtocolBase, TProtocolException
from thrift.Thrift import TApplicationException, TMessageType
from thrift.transport.THeaderTransport import CLIENT_TYPE, THeaderTransport


class THeaderProtocolAccelerate:

    """Base class for pass through header protocols"""

    def get_protocol_id(self):
        raise NotImplementedError()


class THeaderProtocol(TProtocolBase, THeaderProtocolAccelerate):

    """Pass through header protocol (transport can set)"""

    T_BINARY_PROTOCOL = 0
    T_JSON_PROTOCOL = 1
    T_COMPACT_PROTOCOL = 2

    __proto = None
    __proto_id = None

    def get_protocol_id(self):
        return self.__proto_id

    def reset_protocol(self):
        if self.__proto_id == self.trans.get_protocol_id():
            return

        proto_id = self.trans.get_protocol_id()

        if proto_id == self.T_BINARY_PROTOCOL:
            self.__proto = TBinaryProtocolAccelerated(self.trans, self.strictRead, True)
        elif proto_id == self.T_COMPACT_PROTOCOL:
            self.__proto = TCompactProtocolAccelerated(self.trans)
        else:
            raise TApplicationException(
                TProtocolException.INVALID_PROTOCOL, "Unknown protocol requested"
            )
        self.__proto_id = proto_id

    def __init__(self, trans, strictRead=False, client_types=None, client_type=None):
        """Create a THeaderProtocol instance

        @param transport(TTransport) The underlying transport.
        @param strictRead(bool) Turn on strictRead if using TBinaryProtocol
        @param client_types([CLIENT_TYPE.HEADER, ...])
                   List of client types to support.  Defaults to
                   CLIENT_TYPE.HEADER only.
        """

        if isinstance(trans, THeaderTransport):
            trans._THeaderTransport__supported_client_types = set(
                client_types or (CLIENT_TYPE.HEADER,)
            )
            if client_type is not None:
                trans._THeaderTransport__client_type = client_type
            htrans = trans
        else:
            htrans = THeaderTransport(trans, client_types, client_type)
        TProtocolBase.__init__(self, htrans)
        self.strictRead = strictRead
        self.reset_protocol()

    def writeMessageBegin(self, name, type, seqid):
        self.__proto.writeMessageBegin(name, type, seqid)
        if type == TMessageType.CALL or type == TMessageType.ONEWAY:
            # All client to server coms should have a unique seq_id in HEADER
            self.trans.seq_id = seqid

    def writeMessageEnd(self):
        self.__proto.writeMessageEnd()

    def writeStructBegin(self, name):
        self.__proto.writeStructBegin(name)

    def writeStructEnd(self):
        self.__proto.writeStructEnd()

    def writeFieldBegin(self, name, type, id):
        self.__proto.writeFieldBegin(name, type, id)

    def writeFieldEnd(self):
        self.__proto.writeFieldEnd()

    def writeFieldStop(self):
        self.__proto.writeFieldStop()

    def writeMapBegin(self, ktype, vtype, size):
        self.__proto.writeMapBegin(ktype, vtype, size)

    def writeMapEnd(self):
        self.__proto.writeMapEnd()

    def writeListBegin(self, etype, size):
        self.__proto.writeListBegin(etype, size)

    def writeListEnd(self):
        self.__proto.writeListEnd()

    def writeSetBegin(self, etype, size):
        self.__proto.writeSetBegin(etype, size)

    def writeSetEnd(self):
        self.__proto.writeSetEnd()

    def writeBool(self, bool):
        self.__proto.writeBool(bool)

    def writeByte(self, byte):
        self.__proto.writeByte(byte)

    def writeI16(self, i16):
        self.__proto.writeI16(i16)

    def writeI32(self, i32):
        self.__proto.writeI32(i32)

    def writeI64(self, i64):
        self.__proto.writeI64(i64)

    def writeDouble(self, dub):
        self.__proto.writeDouble(dub)

    def writeFloat(self, flt):
        self.__proto.writeFloat(flt)

    def writeString(self, str):
        self.__proto.writeString(str)

    def readMessageBegin(self):
        # Read the next frame, and change protocols if needed
        try:
            self.trans._reset_protocol()
            self.reset_protocol()
        except TApplicationException as ex:
            if self.__proto:
                self.writeMessageBegin(b"", TMessageType.EXCEPTION, 0)
                ex.write(self)
                self.writeMessageEnd()
                self.trans.flush()
        return self.__proto.readMessageBegin()

    def readMessageEnd(self):
        return self.__proto.readMessageEnd()

    def readStructBegin(self):
        return self.__proto.readStructBegin()

    def readStructEnd(self):
        return self.__proto.readStructEnd()

    def readFieldBegin(self):
        return self.__proto.readFieldBegin()

    def readFieldEnd(self):
        return self.__proto.readFieldEnd()

    def readMapBegin(self):
        return self.__proto.readMapBegin()

    def readMapEnd(self):
        return self.__proto.readMapEnd()

    def readListBegin(self):
        return self.__proto.readListBegin()

    def readListEnd(self):
        return self.__proto.readListEnd()

    def readSetBegin(self):
        return self.__proto.readSetBegin()

    def readSetEnd(self):
        return self.__proto.readSetEnd()

    def readBool(self):
        return self.__proto.readBool()

    def readByte(self):
        return self.__proto.readByte()

    def readI16(self):
        return self.__proto.readI16()

    def readI32(self):
        return self.__proto.readI32()

    def readI64(self):
        return self.__proto.readI64()

    def readDouble(self):
        return self.__proto.readDouble()

    def readFloat(self):
        return self.__proto.readFloat()

    def readString(self):
        return self.__proto.readString()


class THeaderProtocolFactory:
    def __init__(self, strictRead=False, client_types=None, client_type=None):
        self.strictRead = strictRead
        self.client_types = client_types
        self.client_type = client_type

    def getProtocol(self, trans):
        prot = THeaderProtocol(
            trans, self.strictRead, self.client_types, self.client_type
        )
        return prot
